pkg/watchdog/watchdog.go

```diff
@@ -0,0 +1,253 @@
+
+	return diskutils.RemoveFile(file)
+}
+
+func WatchdogFileUpdate(watchdogFile string) error {
+	f, err := os.Create(watchdogFile)
+	if err != nil {
+		return err
+	}
+	f.Close()
+
+	return nil
+}
+
+func WatchdogFileExists(baseDir string, vm *v1.VirtualMachine) (bool, error) {
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+
+	filePath := WatchdogFileFromNamespaceName(baseDir, namespace, domain)
+	exists, err := diskutils.FileExists(filePath)
+	if err != nil {
+		return false, err
+	}
+	return exists, nil
+}
+
+func WatchdogFileIsExpired(timeoutSeconds int, baseDir string, vm *v1.VirtualMachine) (bool, error) {
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+
+	filePath := WatchdogFileFromNamespaceName(baseDir, namespace, domain)
+
+	exists, err := diskutils.FileExists(filePath)
+	if err != nil {
+		return false, err
+	}
+
+	if exists == false {
+		return true, nil
+	}
+
+	stat, err := os.Stat(filePath)
+	if err != nil {
+		return false, err
+	}
+
+	now := time.Now().UTC().Unix()
+
+	return isExpired(now, timeoutSeconds, stat), nil
+}
+
+func isExpired(now int64, timeoutSeconds int, stat os.FileInfo) bool {
+	mod := stat.ModTime().UTC().Unix()
+	diff := now - mod
+
+	if diff > int64(timeoutSeconds) {
+		return true
+	}
+	return false
+}
+
+func detectExpiredFiles(timeoutSeconds int, fileDir string) ([]string, error) {
+	var expiredFiles []string
+	files, err := ioutil.ReadDir(fileDir)
+	if err != nil {
+		return nil, err
+	}
+	now := time.Now().UTC().Unix()
+	for _, file := range files {
+		if isExpired(now, timeoutSeconds, file) == true {
+			expiredFiles = append(expiredFiles, file.Name())
+		}
+	}
+	return expiredFiles, nil
+}
+
+type WatchdogListWatcher struct {
+	lock                     sync.Mutex
+	wg                       sync.WaitGroup
+	fileDir                  string
+	stopChan                 chan struct{}
+	eventChan                chan watch.Event
+	watchDogTicker           <-chan time.Time
+	backgroundWatcherStarted bool
+	watchdogTimeout          int
+}
+
+func splitFileNamespaceName(fullPath string) (namespace string, domain string, err error) {
+	fileName := filepath.Base(fullPath)
+	namespaceName := strings.Split(fileName, "_")
+	if len(namespaceName) != 2 {
+		return "", "", fmt.Errorf("Invalid file path: %s", fullPath)
+	}
+
+	namespace = namespaceName[0]
+	domain = namespaceName[1]
+	return namespace, domain, nil
+}
+
+func (d *WatchdogListWatcher) startBackground() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.backgroundWatcherStarted == true {
+		return nil
+	}
+
+	d.stopChan = make(chan struct{}, 1)
+	d.eventChan = make(chan watch.Event, 100)
+
+	tickRate := 1
+
+	if d.watchdogTimeout > 1 {
+		tickRate = d.watchdogTimeout / 2
+	}
+	d.watchDogTicker = time.NewTicker(time.Duration(tickRate) * time.Second).C
+
+	d.wg.Add(1)
```

Comment:
If there is only one go-routine, I normally see that people just use a channel.

Reply:
It doesn't really matter to me.

My thought was that a wg makes it obvious that we're waiting for a goroutine to exit.

Reply:
Fine for me too if you keep it.